### ISSUE TITLE
Adding any code that enables SW interrupts makes the ELF grow huge in size (64M)

### ISSUE THREAD

**Issue by atharvalele** [chipsalliance/riscv-fw-infrastructure#4]
Hi, I was hoping to get some help with the issue referenced below.

Crossposting it here as well for more visibility to the concerned team

https://github.com/platformio/platform-chipsalliance/issues/10

Hope you can help, thank you!

**Comment by oferShinaar** [comment]
Hi @atharvalele ,
I assume you used PIO code, correct? if that is the case, it will be tricky (we are not the owner)
But can you provide the liker script? And .map file... or better share your branch.
Maybe we can have a look and understand what went wrong.

**Comment by atharvalele** [comment]
Hi,

Yes, I'm using PIO. I also forgot to mention that I'm using the Nexys A7 board with the WD EH1 core.

This is the linker script used by PIO: https://pastebin.com/raw/8cSjvbA7

This is my branch with the code: https://github.com/atharvalele/ece540_ece558_pong/tree/board_bringup/fpga_srcs/C_srcs/pong

I seem to have temporarily fixed the issue by commenting out what I thought were the offending lines (L114 - L128). Everything seems to work correctly, but I'm not sure if commenting/removing those lines from the linker script is the right solution.

Thanks for the response!

**Comment by oferShinaar** [comment]
Hi,
The linker script link is broken.
But I can see the one here: https://github.com/atharvalele/ece540_ece558_pong/blob/board_bringup/fpga_srcs/C_srcs/pong/link.lds

Is that the one?

Also, we will need the .map file. If you don’t have it, just create by linker flag: -Wl,-Map=mapfilename.map

**Comment by nati-rapaport** [comment]
@atharvalele the fix you did seems to be ok, however the code set you are using is quite old. You can get the updated one from here - https://github.com/chipsalliance/riscv-fw-infrastructure , and there indeed the linker script does not contain the section that you removed.

**Comment by atharvalele** [comment]
@oferShinaar 

Here's another link for the Linker Script provided by PIO: https://www.toptal.com/developers/hastebin/uqakotikej.yaml

The map file generated by my code is: https://www.toptal.com/developers/hastebin/qedapiwozu.yaml

Let me know if you can access these pastebins.

**Comment by atharvalele** [comment]
@nati-rapaport 

Thanks for the link!

I tried to use the linker script from the repo you linked. But that also generates a binary that is 64M in size.

Here is the .map file for that: https://www.toptal.com/developers/hastebin/ogezuhazud.yaml

I checked the script and I see that some data is placed/copied from `ram` to `ram2` which starts at address 64M. L118 - L127 of link.lds for Nexys A7 EH1 in the master branch. Maybe this is the reason why?

I'm not sure what `ram2` is and what it is being used for.

Thank you so much for the help!

**Comment by oferShinaar** [comment]
@atharvalele the linker script you provide is not from PIO it is from our repo.
After looking on the map file, the 64MB is created because the ELF has to בontinuous and the SW interrupt symbol start from 0x4000000 (that is 64MB)

> @oferShinaar
> 
> Here's another link for the Linker Script provided by PIO: https://www.toptal.com/developers/hastebin/uqakotikej.yaml
> 
> The map file generated by my code is: https://www.toptal.com/developers/hastebin/qedapiwozu.yaml
> 
> Let me know if you can access these pastebins.

@atharvalele the linker script you provide is not from PIO it is from our repo.
After looking at the map file, the 64MB is created because the ELF has to continuous, and the **SW interrupt** symbol start from 0x4000000 (that is 64MB)

**Comment by atharvalele** [comment]
@oferShinaar 

I think then PIO clones your repo to get the linker script.

I'm not sure what the purpose of `ram2` is, could you please provide some details on that?

If I comment out `ram2` related parts from the linker script, everything (including interrupts) still works. Thank you!

**Comment by oferShinaar** [comment]
> 



> @nati-rapaport
> 
> Thanks for the link!
> 
> I tried to use the linker script from the repo you linked. But that also generates a binary that is 64M in size.
> 
> Here is the .map file for that: https://www.toptal.com/developers/hastebin/ogezuhazud.yaml
> 
> I checked the script and I see that some data is placed/copied from `ram` to `ram2` which starts at address 64M. L118 - L127 of link.lds for Nexys A7 EH1 in the master branch. Maybe this is the reason why?
> 
> I'm not sure what `ram2` is and what it is being used for.
> 
> Thank you so much for the help!

```ram2``` is just a logic supporter of 128MB memory on SweRVolf

I don't see where this is coming from:
```
LOAD .pio/build/swervolf_nexys/libBoardBSP.a
LOAD .pio/build/swervolf_nexys/libPSP.a
```
Could be PIO build system?  the ```ld``` file came from our repo and we did not load libs to memory. I think the PIO build system took our libs and put them there.
Check line 267 I believe you will have the same 64MB hole there. We do as well.
After EH1 we started using ```dccm``` instead of ```ram2``` for data. The address of this memory is much closer so the elf will be smaller overall.

Move the PSP section to dccm instead of ram2 and your code will run faster and the elf will be smaller (small gap)
``` 
  PSP_DATA_SEC : ALIGN(16)
  {
    *(.psp_data_section)
    . = ALIGN(4);
  }  > dccm : dccm_load
```

Bottom line, PIO took pieces of our code and mix, so it's hard to track. @nati-rapaport is right. Comment out the lines you did is ok. The linker will find the sections for ```PSP_DATA_SEC ``` in the data area

**Comment by atharvalele** [comment]
@oferShinaar 

I didn't understand why PIO mixed things, it was pretty confusing.

I'll try moving the PSP section to `dccm`.

I think I can close the issue now. Thank you so much for the quick help!
